app/questionnaire/hub.py

```diff
@@ -0,0 +1,136 @@
+            'title': _('Choose another section to complete'),
+            'description': _(
+                'You must complete all sections in order to submit this survey'
+            ),
+            'button': _('Continue'),
+        },
+    }
+
+    COMPLETION_STATES = {
+        'COMPLETED': 'COMPLETED',
+        'IN_PROGRESS': 'IN_PROGRESS',
+        'NOT_STARTED': 'NOT_STARTED',
+    }
+
+    def __init__(self, router, sections):
+        self.router = router
+        self.sections = sections
+
+    def get_content(self, content):
+        return self.HUB_CONTENT_STATES[self.get_survey_state()][content]
+
+    def get_contents(self):
+        return {
+            'title': self.get_content('title'),
+            'description': self.get_content('description'),
+            'rows': self.get_rows(),
+            'submit_button': self.get_content('button'),
+        }
+
+    @staticmethod
+    def get_section_title(section):
+        # Need to handle repeating sections
+        return section['title']
+
+    def get_survey_state(self):
+        if self.router.is_survey_complete():
+            return self.COMPLETION_STATES['COMPLETED']
+
+        return self.COMPLETION_STATES['IN_PROGRESS']
+
+    def get_section_state(self, routing_path, section_id):
+        if self.router.is_section_complete(section_id):
+            return self.COMPLETION_STATES['COMPLETED']
+
+        if self.router.is_section_not_started(routing_path):
+            return self.COMPLETION_STATES['NOT_STARTED']
+
+        return self.COMPLETION_STATES['IN_PROGRESS']
+
+    def get_next_location_for_section(self, state, routing_path):
+        if state == self.COMPLETION_STATES['COMPLETED']:
+            return routing_path[-1]
+
+        if state == self.COMPLETION_STATES['NOT_STARTED']:
+            return self.router.get_first_incomplete_location(routing_path)
+
+        return self.router.get_last_complete_location(routing_path)
+
+    def get_row_for_section(self, section_name, state, section_url):
+        context = {
+            'title': section_name,
+            'rowItems': [
+                {
+                    'valueList': [{'text': self.SECTION_CONTENT_STATES[state]['text']}],
+                    'actions': [
+                        {
+                            'text': self.SECTION_CONTENT_STATES[state]['link']['text'],
+                            'ariaLabel': self.SECTION_CONTENT_STATES[state]['link'][
+                                'aria_label'
+                            ].format(section_name=section_name),
+                            'url': section_url,
+                        }
+                    ],
+                }
+            ],
+        }
+
+        if state == self.COMPLETION_STATES['COMPLETED']:
+            context['rowItems'][0]['icon'] = 'check-green'
+
+        return context
+
+    def get_rows(self):
+        rows = []
+
+        for section in self.sections:
+            routing_path = path_finder.routing_path(section)
```

Comment:
We don't want to be generating a routing path for every section on every visit to the hub. There is no need if the section is complete (the section id will be in completed sections). For in-progress sections we should consider storing that state when a a section is started.

Reply:
Or when a section is revisited

Reply:
@ajmaddaford As per the latest changes, this should be addressed.

@JamesGardiner when a section is revised, the path would need to be evaluated as we can't guarantee something has not changed since the last request, hence we calculate the path on every request?